### PR TITLE
Adding grid layout using phosphor

### PR DESCRIPTION
**Pull request by almarklein**
This is a WIP/stub PR that adds a grid layout based on Phosphor's GridPanel. Supersedes #2863.

We still have the following issues to tackle:
- ~~The content inside the widget is not centered, probably because the plot tries to position itself in a particular way. `margin: auto` does not help here. `text-align: center` centers the plot but not the toolbar. Ideally, the plot should just fill the available space.~~
- ~~The grid panel is now set to fill the entire window, which is a bit of a hack; it works when it is the toplevel widget, but when you nest a grid in another layout, it will break.~~
- (minor issue) there are scrollbars visible, even thought the content should exactly fill the window (probably a stray padding or something).

cc @bryevdv @sccolbert

screenshot from 2015-09-17 17 08 49

### CODE REVIEW

File: bokehjs/gulp/tasks/scripts.coffee [hunk 1]
```diff
@@ -17,8 +17,9 @@ gulp.task "scripts:build", ->
     entries: ['./src/coffee/main.coffee']
     extensions: [".coffee", ".eco"]
     debug: true
-
+  
   browserify opts
+    .transform("browserify-css", {'global': true, 'minify':true})
```
Comment (sccolbert) [inline bokehjs/gulp/tasks/scripts.coffee]: You shouldn't need the global transform anymore, provided you are using the most recent releases of phosphor. Each repo now declares its own required transforms in `package.json`.

File: bokehjs/src/coffee/common/pgrid.coffee [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    window.box = this
+    @views = {}
+    
+    @panel = new gridpanel.GridPanel()
+    @panel.node.style.width = '100%'
+    @panel.node.style.height = '100%'
+    @panel.node.style.position = 'absolute'
+    @el.appendChild(@panel.node)
+    # simulate widget.attachWidget(@panel, @el), but allow @el to be unbound to DOM
+    messaging.sendMessage(@panel, widget.MSG_AFTER_ATTACH)
```
Comment (sccolbert) [inline bokehjs/src/coffee/common/pgrid.coffee]: This is a lie. You should only send this message after the panel is *really* attached to the DOM. Widgets rely on that message being truth.

File: bokehjs/src/coffee/common/pgrid.coffee [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    # Get views
+    child_models = []
+    for row in @mget('children')
+      for child in row
+        if not child?
+          continue
+        child_models.push(child)
+    build_views(@views, child_models, {})
+    # Populate grid
+    @panel.clearChildren()
```
Comment (sccolbert) [inline bokehjs/src/coffee/common/pgrid.coffee]: You don't need to clear children here. It will be done automatically when you assign the children below.

**Comment by sccolbert** [comment]
> The content inside the widget is not centered, probably because the plot tries to position itself in a particular way. margin: auto does not help here.

http://learnlayout.com/margin-auto.html

> The grid panel is now set to fill the entire window, which is a bit of a hack; it works when it is the toplevel widget, but when you nest a grid in another layout, it will break.

That's because of [this](https://github.com/bokeh/bokeh/pull/2871/files#diff-47c18b93edddbee37f74feb7057dbfe5R18) right?

File: bokehjs/src/coffee/common/pgrid.coffee (hunk 2 above)
Comment (almarklein) [inline bokehjs/src/coffee/common/pgrid.coffee]: It works though :)  ok, we'll have to see what is a better moment to send that message.

File: bokehjs/src/coffee/common/pgrid.coffee [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+  initialize: (options) ->
+    super(options)
+    window.box = this
+    @views = {}
+    
+    @panel = new gridpanel.GridPanel()
+    @panel.node.style.width = '100%'
+    @panel.node.style.height = '100%'
+    @panel.node.style.position = 'absolute'
+    @el.appendChild(@panel.node)
```
Comment (almarklein) [inline bokehjs/src/coffee/common/pgrid.coffee]: @sccolbert This is where the widget is set to fill the screen. The widget is put in the correct position in the DOM, but we just make it to fill the whole space, which would not be right if its not the toplevel widget/layout. I consider this a side-effect of using phosphor for one particular widget. If we go all phosphor, the panel would be attached to its parent phosphor panel (or to the root DOM element).

**Comment by almarklein** [comment]
> http://learnlayout.com/margin-auto.html

I don't follow. I tried that, but it does not work.

> That's because of this right? 

No, the `window.box = this` is a leftover from debugging (to provide easy access to the model and panel from the developer console).

File: bokehjs/src/coffee/common/pgrid.coffee (hunk 4 above)
Comment (sccolbert) [inline bokehjs/src/coffee/common/pgrid.coffee]: But the panel node is attached to `PGridView`'s element, so it's only going to fill that element (assuming that element is `position: absolute | relative`).

**Comment by sccolbert** [comment]
> I don't follow. I tried that, but it does not work.

Do you have something I can easily install and run, to take a look?

**Comment by sccolbert** [comment]
From the chat: This is how auto-margin centering works. The explicit width on the div is important:

screen shot 2015-09-17 at 4 30 35 pm

File: bokehjs/src/coffee/common/pgrid.coffee (hunk 4 above)
Comment (almarklein) [inline bokehjs/src/coffee/common/pgrid.coffee]: Oh, right. I just tested, and it seems to work fine when combined with a hplot. Sweet!

**Comment by almarklein** [comment]
Wow, this works better than I'd thought. @bryevdv do you want to have this in 0.10? Remaining points:

* The scrollbars can be removed by setting the body's margin to 0. But maybe this should be part of the plan to move to full phoshphor, and we can live with the scrollbars for now.
* We probably want to use a different name for this (instead of `pgrid`).

**Comment by birdsarah** [comment]
I'm going to give this a test with the responsive.

**Comment by almarklein** [comment]
Tomorrow I don't have much time to work on it except perhaps a tweak in the evening. If this is to land in master tomorrow, someone else should probably wrap it up.

**Comment by birdsarah** [comment]
This almost works with the responsive plots. I just can't figure out how to resize the height of the grid boxes (p-Widgets).

**Comment by almarklein** [comment]
@birdsarah I think you cant, since they're sized by the grid. Can you not set the size of the div *inside* the p-Widget? Better talk to @sccolbert for ideas.

**Comment by birdsarah** [comment]
got to have a lot of good back and forth with @sccolbert last night and am learning the different ways this would/could work. there are a surprising number of variations on a theme that I hadn't thought about and we're going to need some good naming & clarity on these.

**Comment by damianavila** [comment]
This one has been inactive for more than one month as WIP. Closing according to BEP-1. I will let the branch live for a little bit if you want to reopen this one with new work.

**Comment by almarklein** [comment]
Agreed. Please leave the branch; some of this might be useful once we go the Phosphor route with our layour and widgets.

**Pull request closed without merging**